Common/src/main/groovy/io/github/groovymc/cgl/api/datagen/recipe/GShapedRecipeBuilder.groovy

```diff
@@ -0,0 +1,185 @@
+/*
+ * Copyright (C) 2022 GroovyMC and contributors
+ * SPDX-License-Identifier: LGPL-3.0-or-later
+ */
+
+package io.github.groovymc.cgl.api.datagen.recipe
+
+import com.google.common.collect.Lists
+import com.google.common.collect.Maps
+import com.google.common.collect.Sets
+import com.google.gson.JsonArray
+import com.google.gson.JsonObject
+import groovy.transform.CompileStatic
+import net.minecraft.advancements.Advancement
+import net.minecraft.advancements.AdvancementRewards
+import net.minecraft.advancements.RequirementsStrategy
+import net.minecraft.advancements.critereon.RecipeUnlockedTrigger
+import net.minecraft.core.registries.BuiltInRegistries
+import net.minecraft.data.recipes.CraftingRecipeBuilder
+import net.minecraft.data.recipes.FinishedRecipe
+import net.minecraft.resources.ResourceLocation
+import net.minecraft.tags.TagKey
+import net.minecraft.world.item.Item
+import net.minecraft.world.item.ItemStack
+import net.minecraft.world.item.crafting.CraftingBookCategory
+import net.minecraft.world.item.crafting.Ingredient
+import net.minecraft.world.item.crafting.RecipeSerializer
+import net.minecraft.world.level.ItemLike
+import org.jetbrains.annotations.Nullable
+
+import java.util.function.Consumer
+
+@CompileStatic
+class GShapedRecipeBuilder extends CraftingRecipeBuilder implements SimpleRecipeBuilder<GShapedRecipeBuilder> {
+    private final List<String> rows = Lists.newArrayList()
+    private final Map<Character, Ingredient> key = Maps.newLinkedHashMap()
+    private boolean showNotification = true
+
+    GShapedRecipeBuilder define(String symbol, TagKey<Item> tag) {
+        this.define(symbol, Ingredient.of(tag))
+    }
+
+    GShapedRecipeBuilder define(String symbol, ItemLike item) {
+        this.define(symbol, Ingredient.of(item))
+    }
+
+    GShapedRecipeBuilder define(String symbol, Ingredient ingredient) {
+        if (this.key.containsKey(symbol)) {
+            throw new IllegalArgumentException("Symbol '$symbol' is already defined!")
+        } else if (symbol == ' ') {
+            throw new IllegalArgumentException("Symbol ' ' (whitespace) is reserved and cannot be defined")
+        } else {
+            this.key.put(symbol.charAt(0), ingredient)
+            return this
+        }
+    }
+
+    GShapedRecipeBuilder pattern(String... patterns) {
+        patterns.each { pattern ->
+            if (!this.rows.isEmpty() && pattern.length() !== this.rows[0].length()) {
+                throw new IllegalArgumentException('Pattern must be the same width on every line!')
+            } else {
+                this.rows.add(pattern)
+            }
+        }
+        return this
+    }
+
+    GShapedRecipeBuilder showNotification(boolean showNotification) {
```

Comment:
I've seen a few things like this with both a `whatever` and a `setWhatever`; might be worth going through and making sure that if the behavior is the same, one delegates to the other.